src/NVMLIOGroup.cpp

```diff
@@ -0,0 +1,702 @@
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        int result = -1;
+        bool is_found = false;
+        std::shared_ptr<signal_s> signal = m_signal_available.at(signal_name).signals.at(domain_idx);
+
+        // Check if signal was already pushed
+        for (size_t ii = 0; !is_found && ii < m_signal_pushed.size(); ++ii) {
+            // same location means this signal or its alias was already pushed
+            if (m_signal_pushed[ii].get() == signal.get()) {
+                result = ii;
+                is_found = true;
+            }
+        }
+        if (!is_found) {
+            // If not pushed, add to pushed signals and configure for batch reads
+            result = m_signal_pushed.size();
+            signal->m_do_read = true;
+            m_signal_pushed.push_back(signal);
+        }
+
+        return result;
+    }
+
+    // Mark the given control to be written by write_batch()
+    int NVMLIOGroup::push_control(const std::string &control_name, int domain_type, int domain_idx)
+    {
+        if (!is_valid_control(control_name)) {
+            throw Exception("NVMLIOGroup::" + std::string(__func__) + ": control_name " + control_name +
+                            " not valid for NVMLIOGroup",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        if (domain_type != control_domain_type(control_name)) {
+            throw Exception("NVMLIOGroup::" + std::string(__func__) + ": " + control_name + ": domain_type must be " +
+                            std::to_string(control_domain_type(control_name)),
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        if (domain_idx < 0 || domain_idx >= m_platform_topo.num_domain(domain_type)) {
+            throw Exception("NVMLIOGroup::" + std::string(__func__) + ": domain_idx out of range.",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        int result = -1;
+        bool is_found = false;
+        std::shared_ptr<control_s> control = m_control_available.at(control_name).controls.at(domain_idx);
+
+        // Check if control was already pushed
+        for (size_t ii = 0; !is_found && ii < m_control_pushed.size(); ++ii) {
+            // same location means this control or its alias was already pushed
+            if (m_control_pushed[ii] == control) {
+                result = ii;
+                is_found = true;
+            }
+        }
+        if (!is_found) {
+            // If not pushed, add to pushed control
+            result = m_control_pushed.size();
+            m_control_pushed.push_back(control);
+        }
+
+        return result;
+    }
+
+    // Parse and update saved values for signals
+    void NVMLIOGroup::read_batch(void)
+    {
+        m_is_batch_read = true;
+        for (auto &sv : m_signal_available) {
+            if (sv.first == "NVML::CPU_ACCELERATOR_ACTIVE_AFFINITIZATION") {
```

Comment:
This (and the read_signal version) can be broken into helper functions, but I don't believe we should have read_batch call the read_signal version.

Splitting the read_batch and read_signal implementation was an attempt at optimizing the read_batch case.  The underlying NVML call for this signal is nvmlDeviceGetComputeRunningProcesses, which can be costly (it was clocked at 0.5 to 2ms per call, and worst case we call it twice per accelerator).  This returns a list of PIDs associated with each Accelerator.  

In the read_batch case we know that we are looking for this info for all CPUs, so we can cache the PID list at the start of read_batch and then find the CPU-Accelerator mapping for all CPUs.  So it should take num accelerators multiplied by ~0.5 to 2ms + map processing time.

In the read_signal case we don't know that we are looking for more than one CPU, or how long it has been between signal reads, so we call the NVMLDevicePool (and thereby nvmldDeviceGetComputeRunningProcesses) every time.  This difference is key since it will take num cpus multiplied by num accelerators multiplied by ~0.5 to 2ms + map processing time to do the same work as read_batch.